### PR TITLE
Applicant can fill out, submit, and step through all blocks in a given program

### LINKED ISSUE #114: User can submit a program form (submitting updates for a block)
blocked by #113 

**Done when**
- Applicant can submit the block form and persist question answers.
- On success applicant is redirected to the next block (if there is one. create a simple stub "you're done!" page if there isn't)

### PR DESCRIPTION
**Pull request by natsid**
### Description
* Implement `ApplicantProgramBlocksController#update`
  * Parse form data
  * Display errors on invalid form data
  * Redirect to next block on valid form data
  * Redirect to program list on program blocks' completion
  * Handle bad requests
* Add a convenience method on `ReadOnlyApplicantProgramService` that takes a map of updates, since that's the format we get from `DynamicForm`
* Fix up `ReadOnlyApplicantProgrmaService#update` to check for program and applicant not found and throw appropriate errors
* Add controller tests
* Add browser tests

### Checklist
- [x] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)
Fixes #114

### CODE REVIEW

File: universal-application-tool-0.0.1/app/controllers/applicant/ApplicantProgramBlocksController.java [hunk 1]
```diff
@@ -59,9 +73,74 @@ public ApplicantProgramBlocksController(
...
+
+    return applicantService
+        .stageAndUpdateIfValid(applicantId, programId, blockId, formData)
+        .thenApplyAsync(
+            (errorAndROApplicantProgramService) -> {
+              if (errorAndROApplicantProgramService.isError()) {
+                errorAndROApplicantProgramService
+                    .getErrors()
+                    .forEach(e -> logger.error("Exception while updating applicant data", e));
+                return badRequest();
```
Comment (ndmckinley) [inline universal-application-tool-0.0.1/app/controllers/applicant/ApplicantProgramBlocksController.java:89]: Will this be displayed to the user?  Probably some placeholder text here is appropriate.

File: universal-application-tool-0.0.1/test/app/ApplicantProgramBrowserTest.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    fillInput("applicant.address.state", "WA");
+    fillInput("applicant.address.zip", "99999");
+
+    browser.$("button", withText("Save and continue")).click();
+
+    // TODO(https://github.com/seattle-uat/universal-application-tool/issues/256): Expect review
+    //  page when it is implemented.
+    // All blocks complete. Back to list of programs.
+    assertThat(browser.$("h1", withText("Programs")).present()).isTrue();
+    assertThat(browser.$("h2", withText("Mock program")).present()).isTrue();
```
Comment (ndmckinley) [inline universal-application-tool-0.0.1/test/app/ApplicantProgramBrowserTest.java:70]: Can you check that this got saved in the DB?

**Review by ndmckinley — APPROVED** [review]

File: universal-application-tool-0.0.1/test/app/ApplicantProgramBrowserTest.java (hunk 2 above)
Comment (natsid) [reply universal-application-tool-0.0.1/test/app/ApplicantProgramBrowserTest.java:70]: Ah - yes! I can click through `Apply` again and see that the data is present. Good idea.

File: universal-application-tool-0.0.1/app/controllers/applicant/ApplicantProgramBlocksController.java (hunk 1 above)
Comment (natsid) [reply universal-application-tool-0.0.1/app/controllers/applicant/ApplicantProgramBlocksController.java:89]: Ah, yeah. I added some text for now. Might want to spiff it up more later?

File: universal-application-tool-0.0.1/app/services/applicant/Block.java [hunk 3]
```diff
@@ -58,7 +58,7 @@ public String getDescription() {
   public boolean hasErrors() {
     if (errorsMemo.isEmpty()) {
       this.errorsMemo =
-          getQuestions().stream().map(ApplicantQuestion::hasErrors).reduce(Boolean::logicalAnd);
+          getQuestions().stream().map(ApplicantQuestion::hasErrors).reduce(Boolean::logicalOr);
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/applicant/Block.java:61]: 👍🏻

**Review by bion — APPROVED** [review]

**Pull request merged**